### DISCUSSION TITLE
Optional Store when scoping on optional state

### DISCUSSION THREAD

**Discussion by fbp93** [pointfreeco/swift-composable-architecture#957] [Q&A]
I need to perform a scope where the state is optional and being able to transform a `Store<State?, Action>` into `Store<State, Action>?`. 
To avoid some manual unwrap, I tried to add some syntactic sugar, here is how I did it:
```swift
extension Store {
  public func optional<Wrapped: Equatable>() -> Store<Wrapped, Action>? where Wrapped? == State {
      var result: Store<Wrapped, Action>?
      _ = self.publisherScope(state: { $0.compactMap { $0 } }).sink(receiveValue: {
        result = $0
      })
      return result
  }
}

// Usage:
self.store.scope(state: \.state, action: AppAction.action).optional()
```
The problem is that solution is not safe and the `publisherScope` is deprecated. 

What about adding a safe API to retrieve an optional store from an optional state?

**Comment by iampatbrown** [comment]
@fbp93 Is this for UIKit? What do you mean by unsafe? 

Here is an alternate version of your extension using `.ifLet`:

```swift
extension Store {
  public func unwrap<Wrapped>() -> Store<Wrapped, Action>? where State == Wrapped? {
    var unwrapped: Store<Wrapped, Action>?
    _ = self.ifLet { unwrapped = $0 }
    return unwrapped
  }
}
```

**Comment by stephencelis** [comment]
Hi @fbp93! This is precisely what [`Store.ifLet`](https://pointfreeco.github.io/swift-composable-architecture/Store/#store.iflet(then:else:)), which unwraps a store on optional state in the trailing closure.

@iampatbrown's `unwrap` could be used in a synchronous context, but might be dangerous to use over time, so prefer `ifLet` if you can.

I'm going to convert this to a discussion since it's a question about library use and we like to keep the issues section focussed on bugs.

**Reply by iampatbrown** [reply]
@stephencelis agreed. I probably should have mentioned that too :P

**Reply by fbp93** [reply]
Thanks a lot!
